packages/babel-parser/src/util/identifier.js

```diff
@@ -21,3 +21,66 @@ export const keywordRelationalOperator = /^in(stanceof)?$/;
 export function isIteratorStart(current: number, next: number): boolean {
   return current === charCodes.atSign && next === charCodes.atSign;
 }
+
+// This is the comprehensive set of JavaScript reserved words
+// If a word is in this set, it could be a reserved word,
+// depending on sourceType/strictMode/binding info. In other words
+// if a word is not in this set, it is not a reserved word under
+// any circumstance.
+const reservedWordLikeSet = new Set([
+  "break",
+  "case",
+  "catch",
+  "continue",
+  "debugger",
+  "default",
+  "do",
+  "else",
+  "finally",
+  "for",
+  "function",
+  "if",
+  "return",
+  "switch",
+  "throw",
+  "try",
+  "var",
+  "const",
+  "while",
+  "with",
+  "new",
+  "this",
+  "super",
+  "class",
+  "extends",
+  "export",
+  "import",
+  "null",
+  "true",
+  "false",
+  "in",
+  "instanceof",
+  "typeof",
+  "void",
+  "delete",
+  // strict
+  "implements",
+  "interface",
+  "let",
+  "package",
+  "private",
+  "protected",
+  "public",
+  "static",
+  "yield",
+  // strictBind
+  "eval",
+  "arguments",
+  // reservedWorkLike
+  "enum",
+  "await",
+]);
+
+export function canBeReservedWord(word: string): boolean {
+  return reservedWordLikeSet.has(word);
```

Comment:
To avoid maintaining another reserved words list, could this be `word === "await" || word === "enum" || isStrictBindReservedWord(word)`?

Reply:
Instead of that long list, why not do something simpleas `x &  ReservedWord` ?

Reply:
The `Set#has` hashes the input string and checks if the hash value is in the table ([source](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/builtins/builtins-collections-gen.cc;l=2145;drc=be4c5173c2566193f22111961e7f528e4edaa18a?q=SetPrototypeHas&sq=&ss=chromium%2Fchromium%2Fsrc)), since the the input string is scanned only during hash key is computed and the collision is rare by definition of hash, checking presence takes 1 scan of input and O(1) of the set size.

The `StringEquals` first checks the length, then the memory address and then fallbacks to byte-by-byte comparison. ([source](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/builtins/builtins-string-gen.cc;l=242;drc=be4c5173c2566193f22111961e7f528e4edaa18a)) If we check `word === "await"` and then `isStrictBindReservedWord(word)`, the `word` is read for 1.1 times (average on the universe distribution of string length), 1 for the hash computation and 1/10 for comparison with `"await"`. `word === "await"` also introduces extra branch conditions.

> To avoid maintaining another reserved words list

I agree it is not ideal to duplicate keywords, WDYT if I add `canBeReservedWord` to `helper-validator-identifier` in Babel 7.15? Since that will be a feature, I can draft a subseqent PR that moves `canBeReservedWord` to the validator package based on this one so we don't have to wait until 7.15.

> why not do something simple as x & ReservedWord ?

@KFlash Can you elaborate? How can `&` achieves searching here?

Reply:
:+1: for deferring it to 7.15

Reply:
@JLHwung  It's possible if you make things complex and add some bitwise masks behind each reserved word token.

Reply:
Can you verify whether using a prototye-less object is faster? My intuition is that Sets are algorithmically fast, but the actual implementation in V8 is slow.

Reply:
@jridgewell It's faster because I "pack" all reserved words into a set of bitwise masks. It also consumes less memory. You are only dealing with series of number.  see Kataw for an example. Same you can do with tokens, keywords, and everything you need. No need for a lookup.

Reply:
@jridgewell I come up with a [benchmark](https://gist.github.com/JLHwung/7ea981a69867ba9419f089fa49e6566e) file comparing these three approaches: Set#has, Object brand checks and (switch length + string equals). 

The benchmark result on Node.js 16.1.0:
```
object brand checks different string: 2059273 ops/sec ±0.42% (0ms)
object brand checks constant string: 31712533 ops/sec ±0.22% (0ms)
set#has different string: 17756450 ops/sec ±1.26% (0ms)
set#has constant string: 38050433 ops/sec ±1.54% (0ms)
string#equals different string: 13789367 ops/sec ±0.44% (0ms)
string#equals constant string: 709703804 ops/sec ±0.42% (0ms)
```

On different string cases, where each input is in different memory address, `Set#has` is 700% faster than prototype-less object brand check. On constant string cases, `Set#has` is 20% faster than prototype-less object brand check.

On different string cases, `Set#has` is 20% faster than (switch length + string equals). The latter significantly outperform `Set#has` when the input is a constant string, which is expected because V8 optimizes out the per-character equality branch of `StringEquals`. However in Babel parser, each identifier name have different address.

@KFlash Yes we can do a single bitwise check if we replace the data structure of token storage and create different token for different keywords. However Babel currently exposes its token storage via `options.tokens` so we have to maintain compatibility here.